### PR TITLE
[DO NOT MERGE OR REVIEW] bazel: test bazel configuration on a few designs

### PR DESCRIPTION
**Pull request by oharboe**
@jeffng-or @maliberty Just FYI

To build gcd and view gcd:

1. configure your artifact server, same as megaboom with .user-bazelrc
2. `bazel run gcd_route  $(pwd)/build` should not build anything locally, just download artifacts
3. `$(pwd)/build/make gui_final`

### CODE REVIEW

**Comment by oharboe** [comment]
@jeffng-or MockArray is failing in global placement because I've done a mistake.

What is it OpenROAD trying to tell me here?

```
bazel run MockArray_place_deps `pwd`/build
build/make global_place_skip_io_issue
```

Generates:

[global-place-error.tar.gz](https://github.com/user-attachments/files/16612800/global-place-error.tar.gz)


```
./run-me-MockArray-asap7-base.sh
```

How is NonPlaceInstsArea calculated? It should not be equal to CoreArea

```
OpenROAD v2.0-15055-gdb7d56e4d 
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
[INFO GPL-0002] DBU: 1000
[INFO GPL-0003] SiteSize: (  0.054  0.270 ) um
[INFO GPL-0004] CoreBBox: (  2.160  2.160 ) ( 356.400 386.640 ) um
[INFO GPL-0006] NumInstances:             12813
[INFO GPL-0007] NumPlaceInstances:          129
[INFO GPL-0008] NumFixedInstances:         6446
[INFO GPL-0009] NumDummyInstances:         6238
[INFO GPL-0010] NumNets:                  19075
[INFO GPL-0011] NumPins:                  34113
[INFO GPL-0012] DieBBox:  (  0.000  0.000 ) ( 358.560 388.800 ) um
[INFO GPL-0013] CoreBBox: (  2.160  2.160 ) ( 356.400 386.640 ) um
[INFO GPL-0016] CoreArea:            136198.195 um^2
[INFO GPL-0017] NonPlaceInstsArea:   136198.195 um^2
[INFO GPL-0018] PlaceInstsArea:          23.372 um^2
[INFO GPL-0019] Util:                inf        %
[INFO GPL-0020] StdInstsArea:            23.372 um^2
[INFO GPL-0021] MacroInstsArea:           0.000 um^2
[ERROR GPL-0301] Utilization inf % exceeds 100%.
Error: global_place_skip_io.tcl, 20 GPL-0301
```

**Comment by maliberty** [comment]
```
  float util = static_cast<float>(placeInstsArea_)
               / (coreArea - nonPlaceInstsArea_) * 100;
```
so I guess `coreArea == nonPlaceInstsArea_`.  Is the core area completely covered with macros or some such?

**Comment by maliberty** [comment]
```
[INFO GPL-0016] CoreArea:            136198.195 um^2
[INFO GPL-0017] NonPlaceInstsArea:   136198.195 um^2
```

**Comment by oharboe** [comment]
Yes, but I dont see the problem. Looks like a normal mock-array... I dont see anything out of place.

**Comment by maliberty** [comment]
> Yes, but I dont see the problem. Looks like a normal mock-array... I dont see anything out of place.

I see the core and die area are hard coded.  Do they match what an ORFS run generates?  Do you have an image of the floorplan with macros placed?

**Comment by oharboe** [comment]
> > Yes, but I dont see the problem. Looks like a normal mock-array... I dont see anything out of place.
> 
> I see the core and die area are hard coded. Do they match what an ORFS run generates?

yes, I used `make print-DIE_AREA`, etc. to check/extract values.

> Do you have an image of the floorplan with macros placed?

image

**Comment by oharboe** [comment]
This is on master. So there's some blockage there that shouldn't be there...

image


Shouldn't be there:

image

Not there on master:

image

**Comment by maliberty** [comment]
I was about to say the same :)

**Comment by oharboe** [comment]
> I was about to say the same :)

Where does it come from? How do I adjust it?

**Comment by maliberty** [comment]
It comes from scripts/placement_blockages.tcl but apparently it works in ORFS so this is some bazel issue.

**Comment by oharboe** [comment]
> placement_blockages.tcl

Smoking gun...

```
$ make DESIGN_CONFIG=designs/asap7/mock-array/config.mk print-MACRO_BLOCKAGE_HALO
[INFO][FLOW] Invoked hierarchical flow.
Block Element needs to be hardened.
MACRO_BLOCKAGE_HALO = 0
$ build/make print-MACRO_BLOCKAGE_HALO
MACRO_BLOCKAGE_HALO = 
```

Mea-culpa, transliteration error, I forgot about the defaults.mk in mock-array. The mock-array setup can study a range of mock-array configurations, not just 64 bit 8x8...

**Comment by oharboe** [comment]
@hovind What is going on here? I would have expected to find yosys-abc... How can this work if yosys-abc isn't alongside yosys???

```
$ bazel run gcd_synth_deps `pwd`/build
INFO: Invocation ID: 26281929-12aa-4e25-882e-019decd1c797
INFO: Analyzed target //:gcd_synth_deps (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:gcd_synth_deps up-to-date:
  bazel-bin/results/asap7/gcd/base/1_synth.short.mk
INFO: Elapsed time: 0.105s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/gcd_synth_deps.sh /home/oyvind/OpenROAD-flow-scripts/flow/build
$ build/make print-YOSYS_CMD
YOSYS_CMD = external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/tools/install/yosys/bin/yosys
$ find build/external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/tools/install/yosys/bin/
build/external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/tools/install/yosys/bin/
build/external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/tools/install/yosys/bin/yosys
$ ls ../tools/install/yosys/bin/
yosys  yosys-abc  yosys-config  yosys-filterlib  yosys-smtbmc  yosys-witness
```

Aha! It isn't using yosys-abc standalone, but abc command statically linked into Yosys.


```
$ build/make synth
[deleted]
%MKB.' external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/tools/install/yosys/bin/yosys -v 3 -c external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/flow/scripts/synth_canonicalize.tcl) 2>&1 | tee .//logs/asap7/gcd/base/1_1_yosys.log
Warning: Tcl_Init() call failed - no such file or directory
[deleted]
17.16. Finished OPT passes. (There is nothing left to do.)
abc -script external/bazel-orfs~~orfs_repositories~docker_orfs/OpenROAD-flow-scripts/flow/scripts/abc_speed.script -liberty .//objects/asap7/gcd/base/lib/merged.lib -constr .//objects/asap7/gcd/base/abc.constr -dont_use *x1p*_ASAP7* -dont_use *xp*_ASAP7* -dont_use SDF* -dont_use ICG* -D 390
[deleted]
End of script. Logfile hash: 99a2a7237b, CPU: user 1.16s system 0.08s, MEM: 137.19 MB peak
Yosys 0.44 (git sha1 UNKNOWN, clang++ 14.0.0-1ubuntu1.1 -fPIC -O3)
Time spent: 31% 2x abc (0 sec), 20% 2x stat (0 sec), ...
Elapsed time: 0:01.85[h:]min:sec. CPU time: user 1.70 sys 0.11 (98%). Peak memory: 141120KB.
mkdir -p .//results/asap7/gcd/base .//logs/asap7/gcd/base .//reports/asap7/gcd/base
cp .//results/asap7/gcd/base/1_1_yosys.v .//results/asap7/gcd/base/1_synth.v
mkdir -p .//results/asap7/gcd/base
cp designs/asap7/gcd/constraint.sdc .//results/asap7/gcd/base/1_synth.sdc
$ which abc
$
```

**Comment by maliberty** [comment]
yosys doesn't like abc afaik.  The abc command just invokes yosys-abc as a subprocess

**Comment by oharboe** [comment]
> yosys doesn't like abc afaik. The abc command just invokes yosys-abc as a subprocess

Actually, abc is apparently statically linked into yosys and it is run in-process. yosys-abc is a standalone app.

The system abc is not used(and would end in tears I presume...), the abc that shipped with Yosys needs to be used and is used.

**Comment by maliberty** [comment]
hmm... the yosys team just told me yesterday how they don't link abc into yosys (due to stability concerns about abc causing yosys to crash).

**Comment by oharboe** [comment]
More to investigate for me then....

File: flow/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+            "PLACE_DENSITY_LB_ADDON": "0.20",
+        },
+        "cts": SWERV_ALL | {
+            "TNS_END_PERCENT": "100",
+        },
+        "route": SWERV_ALL | {
+            "FASTROUTE_TCL": "$(location :swerv-fastroute)",
+        },
+        "final": SWERV_ALL | {
+            "PWR_NETS_VOLTAGEsS": "",
```
Comment (jeffng-or) [inline flow/BUILD.bazel:92]: Should this be PWR_NETS_VOLTAGES (i.e. remove the lower case s). There's another instance on line 187

**Comment by jeffng-or** [comment]
@oharboe, for the "configure your artifact server, same as megaboom with .user-bazelrc" step, just wanted to double check that this means:

1. create .user-bazelrc in the OpenROAD-flow-scripts/flow directory that includes the call to cred_helper.py
2. For cred_helper.py, did you copy it over from your megaboom workspace? I don't see it listed as part of this PR

**Comment by oharboe** [comment]
well spotted! Yes, it should be copied and kept under git control.

Yes, we are using the same gc bucket and configuration.

File: flow/BUILD.bazel (hunk 1 above)
Comment (oharboe) [reply flow/BUILD.bazel:92]: Yes, well spotted!

We dont run the final stage currently but when we do we would notice.

**Comment by oharboe** [comment]
superceeded by #2320

**Pull request closed without merging**